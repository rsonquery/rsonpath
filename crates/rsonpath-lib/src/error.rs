//! Common errors shared across the library.
use std::fmt::{self, Display};
use thiserror::Error;

pub(crate) const FEATURE_REQUEST_URL: &str =
    "https://github.com/V0ldek/rsonpath/issues/new?template=feature_request.md";
pub(crate) const BUG_REPORT_URL: &str = "https://github.com/V0ldek/rsonpath/issues/new?template=bug_report.md";

/// Internal irrecoverable error. These are caused solely
/// by bugs &ndash; broken invariants or assertions in internal logic &ndash;
/// but we return those instead of panicking.
#[derive(Error, Debug)]
pub struct InternalRsonpathError {
    details: &'static str,
    #[source]
    source: Option<InternalErrorSource>,
}

/// Errors in internal depth tracking of execution engines.
#[derive(Error, Debug)]
pub enum DepthError {
    /// The engine's maximum depth limit was exceeded.
    /// The inner [`usize`] indicates that limit.
    #[error("Maximum depth of {0} exceeded.")]
    AboveLimit(usize),
    /// The document has unmatched closing characters
    /// and is malformed.
    #[error("Depth fell below zero.")]
    BelowZero,
}

struct InternalErrorSource(Box<dyn std::error::Error + Send + Sync>);

impl fmt::Debug for InternalErrorSource {
    #[inline(always)]
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(&self.0, f)
    }
}

impl Display for InternalErrorSource {
    #[inline(always)]
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Display::fmt(&self.0, f)
    }
}

impl std::error::Error for InternalErrorSource {
    #[inline(always)]
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        self.0.source()
    }
}

impl InternalRsonpathError {
    #[allow(unused)]
    pub(crate) fn from_expectation(details: &'static str) -> Self {
        Self { details, source: None }
    }

    #[allow(unused)]
    pub(crate) fn from_error<E: std::error::Error + Send + Sync + 'static>(err: E, details: &'static str) -> Self {
        Self {
            details,
            source: Some(InternalErrorSource(Box::new(err))),
        }
    }
}

impl Display for InternalRsonpathError {
    #[inline]
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let details_are_enabled = std::env::var("RUST_BACKTRACE").unwrap_or_default() == "0";
        write!(
            f,
            "an internal error has occurred; this is a bug, please report it at {BUG_REPORT_URL}"
        )?;

        if details_are_enabled {
            writeln!(f, "; the error details follow")?;
            write!(f, "{}", self.details)?;
            if let Some(source) = &self.source {
                write!(f, "; source: {source}")?;
            }
        }

        Ok(())
    }
}

/// Error raised when rsonpath is asked to perform an operation that is currently
/// unsupported. This may be either because the feature is in the works, or
/// because it is not planned to ever be supported.
#[derive(Error, Debug)]
pub struct UnsupportedFeatureError {
    issue: Option<usize>,
    feature: &'static str,
}

impl UnsupportedFeatureError {
    #[must_use]
    #[allow(dead_code)]
    #[inline(always)]
    fn tracked(issue: usize, feature: &'static str) -> Self {
        Self {
            issue: Some(issue),
            feature,
        }
    }

    #[must_use]
    #[inline(always)]
    fn untracked(feature: &'static str) -> Self {
        Self { issue: None, feature }
    }

    /// Large JSON Depths feature &ndash; supporting JSON documents
    /// with nesting depth exceeding 255. Unsupported and not planned.
    #[must_use]
    #[inline(always)]
    pub fn large_json_depths() -> Self {
        Self::untracked("Large JSON Depths")
    }

    /// Large Automaton Queries feature &ndash; supporting queries that
    /// cause compiled DFAs to exceed 256 states. Unsupported and not planned.
    #[must_use]
    #[inline(always)]
    pub fn large_automaton_queries() -> Self {
        Self::untracked("Large Automaton Queries")
    }

    /// Multiple Selector per Segment feature &ndash; supporting queries
    /// that contain a union of selectors per step. Unsupported and not planned (yet).
    #[must_use]
    #[inline(always)]
    pub fn multiple_selectors() -> Self {
        Self::untracked("Multiple Selector per Segment")
    }

    /// Indexing from End &ndash; supporting index and slice selectors that
    /// use from-end indexing. Unsupported and not planned (yet).
    #[must_use]
    #[inline(always)]
    pub fn indexing_from_end() -> Self {
        Self::untracked("Indexing from End")
    }

    /// Slice Selector &ndash; supporting slice selectors.
    /// <https://github.com/V0ldek/rsonpath/issues/152>
    #[must_use]
    #[inline(always)]
    pub fn slice_selector() -> Self {
        Self::tracked(152, "Slice Selector")
    }

    /// Slice with Backward Step &ndash; supporting slice selectors that step backwards.
    /// Unsupported and not planned (yet).
    #[must_use]
    #[inline(always)]
    pub fn slice_with_backward_step() -> Self {
        Self::untracked("Slice with Backward Step")
    }

    /// Filter Selector &ndash; supporting filter selectors.
    /// <https://github.com/V0ldek/rsonpath/issues/154>
    #[must_use]
    #[inline(always)]
    pub fn filter_selector() -> Self {
        Self::tracked(154, "Filter Selector")
    }

    /// Returns the issue number on GitHub corresponding to the unsupported feature.
    /// Is [`None`] if the feature is not planned.
    #[must_use]
    #[inline(always)]
    pub fn issue(&self) -> Option<usize> {
        self.issue
    }

    /// Returns the descriptive name of the feature.
    #[must_use]
    #[inline(always)]
    pub fn feature(&self) -> &str {
        self.feature
    }

    /// Whether the issue is planned to ever be supported.
    #[must_use]
    #[inline(always)]
    pub fn is_planned(&self) -> bool {
        self.issue.is_some()
    }
}

impl Display for UnsupportedFeatureError {
    #[inline]
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self.issue {
            Some(issue) => {
                write!(
                    f,
                    "the feature {} (#{}) is not supported yet; it is being tracked and discussed at \
                    https://github.com/V0ldek/rsonpath/issues/{}",
                    self.feature, issue, issue
                )
            }
            None => {
                write!(
                    f,
                    "the feature {} is not supported, and is not planned; \
                    if you would like to see it introduced to rsonpath, please raise a feature request at \
                    {FEATURE_REQUEST_URL}",
                    self.feature
                )
            }
        }
    }
}
