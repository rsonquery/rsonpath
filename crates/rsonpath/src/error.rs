use color_eyre::{Help, SectionExt};
use eyre::eyre;
use rsonpath_lib::{automaton::error::CompilerError, engine::error::EngineError, error::UnsupportedFeatureError};
use rsonpath_syntax::{error::ParseError, JsonPathQuery};

const FEATURE_REQUEST_URL: &str = "https://github.com/V0ldek/rsonpath/issues/new?template=feature_request.md";

/// Turn a [`ParseError`] into a user-friendly eyre Report.
pub(super) fn report_parser_error(error: ParseError) -> eyre::Report {
    eyre!("One or more syntax errors occurred.").section(error.colored().header("Parse error:"))
}

/// Turn a [`CompilerError`] into a user-friendly eyre Report.
pub(super) fn report_compiler_error(query: &JsonPathQuery, error: CompilerError) -> eyre::Report {
    match error {
        CompilerError::NotSupported(unsupported) => report_unsupported_error(&unsupported),
        CompilerError::QueryTooComplex(_) => {
            let mut report = eyre::Report::new(error);
            if query.segments().iter().any(|segment| {
                segment
                    .selectors()
                    .iter()
                    .any(|selector| matches!(selector, rsonpath_syntax::Selector::Wildcard))
            }) {
                report = report.suggestion(
                    "Wildcard selectors are a common source of query complexity.\n            \
                    Consider reformulating the query using descendant selectors to replace sequences of wildcards.",
                );
            }
            add_unsupported_context(report, &UnsupportedFeatureError::large_automaton_queries())
        }
    }
}

/// Turn a [`EngineError`] into a user-friendly eyre Report.
pub(super) fn report_engine_error(error: EngineError) -> eyre::Report {
    match error {
        EngineError::DepthAboveLimit(_, _) => {
            add_unsupported_context(eyre::Report::new(error), &UnsupportedFeatureError::large_json_depths())
        }
        EngineError::NotSupported(unsupported) => report_unsupported_error(&unsupported),
        EngineError::DepthBelowZero(_, _)
        | EngineError::InputError(_)
        | EngineError::InternalError(_)
        | EngineError::MalformedStringQuotes(_)
        | EngineError::MissingClosingCharacter()
        | EngineError::MissingItem()
        | EngineError::MissingOpeningCharacter()
        | EngineError::SinkError(_) => eyre::Report::new(error),
    }
}

fn report_unsupported_error(unsupported: &UnsupportedFeatureError) -> eyre::Report {
    use color_eyre::owo_colors::OwoColorize;
    let feature = unsupported.feature();
    let base_report = if unsupported.is_planned() {
        let feature = feature.blue();
        eyre!("The feature {feature} {}", "is not supported yet.".bright_red())
    } else {
        let feature = feature.red();
        eyre!("The feature {feature} {}", "is not supported.".bright_red())
    };
    add_unsupported_context(base_report, unsupported)
}

fn add_unsupported_context(report: eyre::Report, unsupported: &UnsupportedFeatureError) -> eyre::Report {
    use color_eyre::owo_colors::OwoColorize;
    let feature = unsupported.feature();
    if let Some(issue) = unsupported.issue() {
        let feature = feature.blue();
        report.note(format!(
            "The feature {feature} is planned for a future release of rsonpath.\n      \
            You can join the ongoing discussion at {}.",
            format!("https://github.com/V0ldek/rsonpath/issues/{issue}").bright_blue()
        ))
    } else {
        let feature = feature.red();
        report.note(format!(
            "The feature {feature} is not supported and is {} planned.\n      \
            If you would like to see it introduced to rsonpath, please raise a feature request at\n      \
            {} and describe your use case.",
            "not".italic(),
            FEATURE_REQUEST_URL.bright_blue()
        ))
    }
}
