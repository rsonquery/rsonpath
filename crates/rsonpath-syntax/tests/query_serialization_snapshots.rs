#[cfg(feature = "serde")]
mod ron {
    use insta::assert_ron_snapshot;
    use rsonpath_syntax::parse;
    use std::error::Error;

    #[test]
    fn empty_query() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&parse("$")?);
        Ok(())
    }

    #[test]
    fn readme_query() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&parse("$.jsonpath[*]")?);
        Ok(())
    }

    #[test]
    fn jsonpath_example_query() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&parse("$..phoneNumbers[*].number")?);
        Ok(())
    }

    #[test]
    fn real_life_query() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&parse("$.personal.details.contact.information.phones.home")?);
        Ok(())
    }

    #[test]
    fn slice() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&parse("$..entries[3:5:7]")?);
        Ok(())
    }

    #[test]
    fn multiple_selectors() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&parse("$..entries['abc', 4, 7:10:13]")?);
        Ok(())
    }

    #[test]
    fn filter() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&parse("$..user[?@.id == 'value']..entities..url")?);
        Ok(())
    }

    #[test]
    fn nested_filter() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&parse(r#"$.a[?@.b == "abc" && $[*][?@.c > 3.13]]"#)?);
        Ok(())
    }
}
