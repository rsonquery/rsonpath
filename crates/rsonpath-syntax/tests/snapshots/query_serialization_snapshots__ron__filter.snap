---
source: crates/rsonpath-syntax/tests/query_serialization_snapshots.rs
expression: "&parse(\"$..user[?@.id == 'value']..entities..url\")?"
snapshot_kind: text
---
JsonPathQuery(
  segments: [
    Descendant(Selectors(
      inner: [
        Name(JsonString(
          quoted: "\"user\"",
        )),
      ],
    )),
    Child(Selectors(
      inner: [
        Filter(Comparison(ComparisonExpr(
          lhs: RelativeSingularQuery(SingularJsonPathQuery(
            segments: [
              Name(JsonString(
                quoted: "\"id\"",
              )),
            ],
          )),
          op: EqualTo,
          rhs: Literal(String(JsonString(
            quoted: "\"value\"",
          ))),
        ))),
      ],
    )),
    Descendant(Selectors(
      inner: [
        Name(JsonString(
          quoted: "\"entities\"",
        )),
      ],
    )),
    Descendant(Selectors(
      inner: [
        Name(JsonString(
          quoted: "\"url\"",
        )),
      ],
    )),
  ],
)
